src/imitation/scripts/train_adversarial.py

```diff
@@ -284,22 +277,32 @@ def add_data_ep_reward(self, epoch):
       # Don't calculate ep reward twice.
       return
     self.ep_reward_X.append(epoch)
-    self._add_data_ep_reward(self.venv_norm_obs, "Ground Truth Reward")
-    self._add_data_ep_reward(self.venv_train_norm_obs, "Train Reward")
-    self._add_data_ep_reward(self.venv_test_norm_obs, "Test Reward")
-
-  def _add_data_ep_reward(self, env, name):
-    sample_until = util.rollout.min_episodes(self.n_episodes_per_reward_data)
 
     gen_policy = self.trainer.gen_policy
-    gen_ret = util.rollout.mean_return(gen_policy, env, sample_until)
-    self.gen_ep_reward[name].append(gen_ret)
-    tf.logging.info("generator return: {}".format(gen_ret))
-
     rand_policy = util.init_rl(self.trainer.venv)
-    rand_ret = util.rollout.mean_return(rand_policy, env, sample_until)
-    self.rand_ep_reward[name].append(rand_ret)
-    tf.logging.info("random return: {}".format(rand_ret))
+    sample_until = util.rollout.min_episodes(self.n_episodes_per_reward_data)
+    trajs_rand_orig_rew = util.rollout.generate_trajectories(
+      rand_policy, self.venv_norm_obs, sample_until)
+    trajs_gen_orig_rew = util.rollout.generate_trajectories(
+      gen_policy, self.venv_norm_obs, sample_until)
+
+    for reward_fn, reward_name in [(None, "Ground Truth Reward"),
+                                   (self.trainer.reward_train, "Train Reward"),
+                                   (self.trainer.reward_test, "Test Reward")]:
+      if reward_fn is None:
+        trajs_rand, trajs_gen = trajs_rand_orig_rew, trajs_gen_orig_rew
+      else:
+        trajs_rand = [util.rollout.rewrite_rewards_traj(t, reward_fn)
+                      for t in trajs_rand_orig_rew]
+        trajs_gen = [util.rollout.rewrite_rewards_traj(t, reward_fn)
+                     for t in trajs_gen_orig_rew]
+
+      gen_ret = util.rollout.rollout_stats(trajs_gen)["return_mean"]
```

Comment:
I'm wondering if we can compute this more directly? If I understand correctly, the reason we need trajectories here is because that's what `rollout_stats` expects, but we only actually need the returns to compute these metrics.

How about modifying `rewrite_reward_trajs` to just return `new_rew` rather than a new `Trajectory`? If we called this `compute_rewards`, we could then compute the mean returns directly:
```
gen_ret = np.mean([sum(compute_rewards(t)) for t in trajs_gen_orig_rew])
```

If we wanted things other than just the mean, then it might make sense to instead factor out the summary statistics generating functionality from `rollout_stats` from constructing the various `traj_descriptors`.

Reply:
Good point, there's no need to regenerate the entire Trajectory and compute every stat. I've modified `rewrite_reward_trajs` to return new rewards only and calculate mean rewards only.